nexus/types/src/deployment/planning_input.rs

```diff
@@ -1043,6 +1044,190 @@ pub struct Policy {
+        }
+
+        Ok(builder.build())
+    }
+}
+
+impl ExternalIpPolicy {
+    pub fn empty() -> Self {
+        Self {
+            service_ip_pool_ranges: Vec::new(),
+            external_dns_ips: BTreeSet::new(),
+        }
+    }
+
+    /// Construct an `ExternalIpPolicy` containing a single service IP pool and
+    /// no external DNS IPs.
+    ///
+    /// This is used primarily by tests. A single pool with no external DNS IPs
+    /// can be constructed infallibly.
+    pub fn single_pool_no_external_dns(range: IpRange) -> Self {
+        Self {
+            service_ip_pool_ranges: vec![range],
+            external_dns_ips: BTreeSet::new(),
+        }
+    }
+
+    /// Construct an empty [`ExternalIpPolicyBuilder`].
+    pub fn builder() -> ExternalIpPolicyBuilder {
+        ExternalIpPolicyBuilder::new()
+    }
+
+    /// Construct an [`ExternalIpPolicyBuilder`] that contains all of this
+    /// policy's IP pools and external DNS IPs.
+    pub fn into_builder(self) -> ExternalIpPolicyBuilder {
+        let Self { service_ip_pool_ranges, external_dns_ips } = self;
+        ExternalIpPolicyBuilder { service_ip_pool_ranges, external_dns_ips }
+    }
+
+    /// Consume this `ExternalIpPolicy`, returning an iterator over all IPs that
+    /// should be used for services other than external DNS (i.e., Nexus and
+    /// boundary NTP).
+    pub fn into_non_external_dns_ips(self) -> impl Iterator<Item = IpAddr> {
+        let Self { service_ip_pool_ranges, external_dns_ips } = self;
+        service_ip_pool_ranges
+            .into_iter()
+            .flat_map(|r| r.iter())
+            .filter(move |ip| !external_dns_ips.contains(ip))
+    }
+
+    /// The set of external IP addresses on which we should run external DNS
+    /// servers.
+    pub fn external_dns_ips(&self) -> &BTreeSet<IpAddr> {
+        &self.external_dns_ips
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct ExternalIpPolicyBuilder {
+    service_ip_pool_ranges: Vec<IpRange>,
+    external_dns_ips: BTreeSet<IpAddr>,
+}
+
+impl ExternalIpPolicyBuilder {
+    /// Create a new, empty `ExternalIpPolicyBuilder`.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Add an external IP pool for services.
+    ///
+    /// # Errors
+    ///
+    /// Fails if `pool_range` overlaps an existing pool within this builder.
+    // TODO-correctness Eventually this may be broken up into different pools
+    // for different kinds of services. For now, we group all services into a
+    // single set of pools and then separately carve out external DNS from
+    // within those pools.
+    pub fn push_service_ip_pool(
+        &mut self,
+        new_pool: IpRange,
+    ) -> Result<(), ExternalIpPolicyError> {
+        // Linear scan should be good enough: we don't expect many pools, and
+        // the check we do for each pool is cheap.
+        for &existing_pool in &self.service_ip_pool_ranges {
+            if existing_pool.overlaps(&new_pool) {
```

Comment:
Do we care about IP versions here? In particular, it seems possible to push an IPv4 range into an IPv6 IP Pool, if I'm understanding the `overlaps()` method and other checks here.

Reply:
That all sounds like a good middle ground to me. I'm on board.

Reply:
Took a stab at this in [c385f6d](https://github.com/oxidecomputer/omicron/pull/9291/commits/c385f6d85a893fbc50b1da4bc9153860a6a30f86) - what do you think?